powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/encryption/EncryptionService.java

```diff
@@ -0,0 +1,245 @@
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.getlime.security.powerauth.app.server.service.encryption;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.Objects;
+
+import javax.crypto.SecretKey;
+
+import org.apache.commons.lang3.ArrayUtils;
+import org.springframework.stereotype.Service;
+
+import io.getlime.security.powerauth.app.server.configuration.PowerAuthServiceConfiguration;
+import io.getlime.security.powerauth.app.server.database.model.enumeration.EncryptionMode;
+import io.getlime.security.powerauth.app.server.service.exceptions.GenericServiceException;
+import io.getlime.security.powerauth.app.server.service.i18n.LocalizationProvider;
+import io.getlime.security.powerauth.app.server.service.model.ServiceError;
+import io.getlime.security.powerauth.crypto.lib.generator.KeyGenerator;
+import io.getlime.security.powerauth.crypto.lib.model.exception.CryptoProviderException;
+import io.getlime.security.powerauth.crypto.lib.model.exception.GenericCryptoException;
+import io.getlime.security.powerauth.crypto.lib.util.AESEncryptionUtils;
+import io.getlime.security.powerauth.crypto.lib.util.KeyConvertor;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Service for encryption and decryption database data.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Service
+@Slf4j
+@AllArgsConstructor
+public class EncryptionService {
+
+    private final PowerAuthServiceConfiguration powerAuthServiceConfiguration;
+    private final LocalizationProvider localizationProvider;
+
+    private final KeyGenerator keyGenerator = new KeyGenerator();
+    private final AESEncryptionUtils aesEncryptionUtils = new AESEncryptionUtils();
+    private final KeyConvertor keyConvertor = new KeyConvertor();
+
+    /**
+     * Decrypt the given data.
+     *
+     * @param data Data to decrypt.
+     * @param encryptionMode Encryption mode.
+     * @param secretKeyDerivationInput Values used for derivation of secret key.
+     * @return Decrypted value.
+     * @throws GenericServiceException In case decryption fails.
+     */
+    public byte[] decrypt(final byte[] data, final EncryptionMode encryptionMode, final List<String> secretKeyDerivationInput) throws GenericServiceException {
```

Comment:
Much saver now. Input and output are a byte array, the encoding type is driven by converters,